lib/spack/spack/solver/asp.py

```diff
@@ -763,6 +770,11 @@ def key_fn(version):
                 )
             )
 
+        for v1 in most_to_least_preferred:
+            for v2 in most_to_least_preferred:
+                if equivalent_versions(v1.version, v2.version):
+                    self.gen.fact(fn.version_equivalent(pkg.name, v1.version, v2.version))
```

Comment:
Is there some issue with

```
for version in most_to_least_preferred:
    if isinstance(version, spack.version.GitVersion):
        self.gen.fact(fn.version_equivalent(pkg.name, version.version, version.ref_version)
        if version.ref_version not in most_to_least_preferred:
            self.gen.fact(fn.version_declared(pkg.name, version.ref_version, ARBITRARY_HIGH_WEIGHT, 'git_ref')
```

and add `'git_ref'` to the `version_origin_fields` list as the last element.

That would seem to be more straightforward and avoid the quadratic iteration.

Reply:
Hmm, I hadn't thought about it like this. I was thinking we were matching the specs, but yeah I guess we are just matching the versions after all.  Isn't this still quadratic (with some optimization) due to this?

`if version.ref_version not in most_to_least_preferred:`

Do we even need this though? if the `ref_version` is not in `most_to_least_preferred` then how could the solver even select it as an equivalent version in the first place?